File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        output = []
        start_row = 0
        end_row = len(matrix)
        start_column = 0
        end_column = len(matrix[0])

        while (start_row < end_row and start_column < end_column):

            # traverse from left to right
            for i in range(start_column, end_column):
                output.append(matrix[start_row][i])
            
            start_row += 1

            # traverse from top to bottom
            for i in range(start_row, end_row):
                output.append(matrix[i][end_column - 1])
            
            end_column -=1

            if (start_row < end_row):
                # traverse from right to left
                for i in range(end_column - 1, start_column - 1, - 1):
                    output.append(matrix[end_row - 1][i])
                end_row -= 1

            if (start_column < end_column):
                # traverse from bottom to top
                for i in range(end_row - 1, start_row - 1, - 1):
                    output.append(matrix[i][start_column])
                start_column += 1

        return output
        

File: 54-spiral-matrix/spiral-matrix.js
// declare a variable output that will be returned
// take the dimension of the matrix
// traverse the matrix in a spiral order while still in bounds

var spiralOrder = function(matrix) {
    let output = []
    let topRow = 0
    let bottomRow = matrix.length - 1
    let leftColumn = 0
    let rightColumn = matrix[0].length - 1

    while (topRow <= bottomRow && leftColumn <= rightColumn) {

        // traverse left to right
        for (let i = leftColumn; i <= rightColumn; i++) {
            output.push(matrix[topRow][i])
        }
        topRow ++


        // traverse top to bottom
        for (let i = topRow; i <= bottomRow; i++) {
            output.push(matrix[i][rightColumn])
        }
        rightColumn --


        if (topRow <= bottomRow) {
            // traverse right to left
            for (let i = rightColumn; i >= leftColumn; i--) {
                output.push(matrix[bottomRow][i])
            }
            bottomRow --

        }

        if (leftColumn <= rightColumn) {
            // traverse bottom to top
            for (let i = bottomRow; i >= topRow; i--) {
                output.push(matrix[i][leftColumn])
            }
            leftColumn++

        }

    }

    return output
        
};